### ISSUE TITLE
Allow renaming of folders

### ISSUE THREAD

**Issue by AlexanderThaller** [ChuckDaniels87/rnr#9]
It would be also usefull to be able to rename folders not just files.

**Comment by ChuckDaniels87** [comment]
I considered it at first, but it is really non-trivial to rename directories. Changing a directory name also affects all files and subdirectories.
I want to include other features first, but feel free to discuss a possible implementation here.

**Comment by AlexanderThaller** [comment]
Yeah I was just thinking about one level of renaming doing it recursively is probably not as trivial.

I don't know how the renaming is implemented right now but probably I would build a tree of the files/folders that need to be renamed and then start with the last leaf or something.

So with a folder stucture like this:

```
/tmp/testfolder
» tree
.
└── test01
    ├── test01_01
    └── test01_02
        ├── test01_02_01
        └── test01_02_02

5 directories, 0 files
```

if would rename them like that (from test to blub)

```
rnr -r test blub .
```

```
1. step: test01/test01_02/test01_02_01 -> test01/test01_02/blub01_02_01
2. step: test01/test01_02/test01_02_02 -> test01/test01_02/blub01_02_02
3. step: test01/test01_01 -> test01/blub01_01
4. step: test01/test01_02 -> test01/blub01_02
5. step: test01 -> blub01
```

Not sure if that makes sense.

**Comment by ChuckDaniels87** [comment]
I think that your idea is actually doable but I think that resources consumption and solving time could increase a lot. In fact, I suspect that other popular renamers do not include directory renaming for this reason among others.

> Maybe it would also be easier to just get the paths and sort them (have to test how sorting of paths is implemented in rust).

We should take a deeper look at this. Renaming conflicts will be much more difficult to solve since complex dependency tree. However, I am using [`walkdir` crate](https://docs.rs/walkdir/2.1.4/walkdir/) that can help with your proposed idea.

In any case, I think directory renaming must be set with an optional flag because it is not what is usually expected from a renamer. Actually, the problem could be separated in renaming directories-only and files-only with two different calls to `rnr`. This could make the problem much simpler.

There is heavy code refactoring and restructuring right now. I am working right now in a more reliable solver to handle the renaming order. I will take a look at this after releasing 'v0.2.0'.

**Comment by ChuckDaniels87** [comment]
Sure, a PR or a separated PoC will be welcomed. Enjoy your vacation!

**Comment by ChuckDaniels87** [comment]
I've found out something that must be taken into account to rename directories. The current implementation uses `std::fs::rename` function, and [its documentation](https://doc.rust-lang.org/std/fs/fn.rename.html#platform-specific-behavior) states the following conditions to move a directory:
> Because of this, the behavior when both from and to exist differs. On Unix, if from is a directory, to must also be an (empty) directory. If from is not a directory, to must also be not a directory. In contrast, on Windows, from can be anything, but to must not be a directory.

**Comment by AlexanderThaller** [comment]
Thank you for implementing this. Sry that I didn't find the time to tackle this.

**Comment by sergeevabc** [comment]
Quite strange there is no flexibility to rename folders only but not files.

**Comment by ismaelgv** [comment]
@sergeevabc feel free to open a new issue for this feature. It could be added as a new flag.
